fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/TableQueue.java

```diff
@@ -0,0 +1,159 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades;
+
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.cursors.ListCursor;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.planprotos.PTableQueue;
+import com.apple.foundationdb.record.query.plan.plans.QueryResult;
+import com.apple.foundationdb.tuple.ByteArrayUtil2;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.ZeroCopyByteString;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+/**
+ * A mutable, temporary, serializable, and in-memory buffer of {@link QueryResult}s. It is aimed to be used as a temporary
+ * placeholder for computation results produced by some physical operator, but can be leveraged to represent, for example, temporary
+ * tables as well.
+ */
+public class TableQueue {
+
+    @Nonnull
+    private final Queue<QueryResult> underlyingBuffer;
+
+    @Nonnull
+    private final CorrelationIdentifier name;
+
+    private TableQueue(@Nonnull CorrelationIdentifier name) {
+        this(new LinkedList<>(), name);
+    }
+
+    private TableQueue(@Nonnull Queue<QueryResult> buffer, @Nonnull CorrelationIdentifier name) {
+        this.underlyingBuffer = buffer;
+        this.name = name;
+    }
+
+    @Nonnull
+    public CorrelationIdentifier getName() {
+        return name;
+    }
+
+    /**
+     * Add a new {@link QueryResult} element to the queue.
+     * @param element the new element to be added.
+     */
+    public void add(@Nonnull QueryResult element) {
+        underlyingBuffer.add(element);
+    }
+
+    /**
+     * Add a new {@link QueryResult} elements to the queue.
+     * @param elements the new elements to be added.
+     */
+    public void add(@Nonnull QueryResult... elements) {
+        Arrays.stream(elements).forEach(this::add);
+    }
+
+
+    @Nonnull
+    public Queue<QueryResult> getReadBuffer() {
+        return underlyingBuffer;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nonnull
+    public RecordCursor<QueryResult> getReadCursor(@Nullable byte[] continuation) {
+        return new ListCursor<>((List<QueryResult>)getReadBuffer(), continuation);
```

Comment:
Because this exhausts, it does not work to have one part of the plan filling and another part emptying. One must, I think, do all the `add`s and _then_ `getReadCursor` to process the queue. Does that mean that the table queue itself ought to have a mode for which direction it is currently handling and callers must set that explicitly?

Reply:
It is intended to be used such that there are no interleaving reads and writes, to address that, the consumer must make sure that these accesses are synchronised correctly e.g. by means of read-write lock, otherwise it would might break any expected continuation semantics imposed by the consumer.